include/nbl/builtin/hlsl/bxdf/fresnel.hlsl

```diff
@@ -17,131 +18,354 @@ namespace hlsl
     {
-        scalar_type NdotT2 = rcpOrientedEta2 * NdotI2 + 1.0 - rcpOrientedEta2;
-        scalar_type absNdotT = sqrt<scalar_type>(NdotT2);
-        return backside ? absNdotT : -(absNdotT);
+        this_t retval;
+        retval.I = I;
+        retval.N = N;
+        retval.NdotI = NdotI;
+        retval.NdotI2 = retval.NdotI * retval.NdotI;
+        retval.rcpOrientedEta = rcpEtas;
+        retval.recomputeNdotT(rcpEtas.backside, retval.NdotI2, rcpEtas.value2);
+        return retval;
     }
 
-    vector_type doRefract()
+    void recomputeNdotI()
     {
-        return N * (NdotI * rcpOrientedEta + computeNdotT(backside, NdotI2, rcpOrientedEta2)) - rcpOrientedEta * I;
+        NdotI = hlsl::dot<vector_type>(N, I);
+        NdotI2 = NdotI * NdotI;
     }
 
-    static vector_type doReflectRefract(bool _refract, NBL_CONST_REF_ARG(vector_type) _I, NBL_CONST_REF_ARG(vector_type) _N, scalar_type _NdotI, scalar_type _NdotTorR, scalar_type _rcpOrientedEta)
+    void recomputeNdotT(bool backside, scalar_type _NdotI2, scalar_type rcpOrientedEta2)
     {
-        return _N * (_NdotI * (_refract ? _rcpOrientedEta : 1.0f) + _NdotTorR) - _I * (_refract ? _rcpOrientedEta : 1.0f);
+        scalar_type NdotT2 = rcpOrientedEta2 * _NdotI2 + 1.0 - rcpOrientedEta2;
+        scalar_type absNdotT = sqrt<scalar_type>(NdotT2);
+        NdotT = ieee754::flipSign(absNdotT, backside);
     }
 
-    vector_type doReflectRefract(bool r)
+    vector_type operator()()
     {
-        const T NdotTorR = r ? computeNdotT(backside, NdotI2, rcpOrientedEta2) : NdotI;
-        return doReflectRefract(r, I, N, NdotI, NdotTorR, rcpOrientedEta);
+        recomputeNdotT(rcpOrientedEta.backside, NdotI2, rcpOrientedEta.value2);
+        return N * (NdotI * rcpOrientedEta.value + NdotT) - rcpOrientedEta.value * I;
     }
 
     vector_type I;
     vector_type N;
-    bool backside;
+    scalar_type NdotT;
     scalar_type NdotI;
     scalar_type NdotI2;
+    fresnel::OrientedEtaRcps<scalar_type> rcpOrientedEta;
+};
+
+template<typename T NBL_PRIMARY_REQUIRES(concepts::Scalar<T>)
+struct ReflectRefract
+{
+    using this_t = ReflectRefract<T>;
+    using vector_type = vector<T, 3>;
+    using scalar_type = T;
+
+    static this_t create(bool r, NBL_CONST_REF_ARG(vector_type) I, NBL_CONST_REF_ARG(vector_type) N, scalar_type NdotI, scalar_type NdotTorR, scalar_type rcpOrientedEta)
+    {
+        this_t retval;
+        retval.I = I;
+        retval.N = N;
+        retval.NdotI = NdotI;
+        retval.NdotTorR = NdotTorR;
+        retval.rcpOrientedEta = rcpOrientedEta;
+        return retval;
+    }
+
+    static this_t create(bool r, NBL_CONST_REF_ARG(Refract<scalar_type>) refract)
+    {
+        this_t retval;
+        retval.I = refract.I;
+        retval.N = refract.N;
+        retval.NdotI = refract.NdotI;
+        retval.NdotTorR = hlsl::mix(refract.NdotI, refract.NdotT, r);
```

Comment:
don't store a `NdotTorR` because that forces me to "sync" `create(bool refract,..)` with `operator()(const bool refract)`

The idea is that I create once, and I call `operator()` changing reflection vs refraction choice at will.

So we can't keep a cached `NdotTorR` member, we need to compute it in `operator()`

You can add `NdotR()` and `NdotT()` members for when one knows it will reflect or refract respectively

Reply:
can accept
```cpp
r(doRefract,r.computeNdotTorR())
```